src/gui/uithememanager.h

```diff
@@ -36,6 +36,16 @@
 #include <QObject>
 #include <QString>
 
+class UIThemeSource
+{
+public:
+    virtual ~UIThemeSource() = default;
+
+    virtual QByteArray styleSheet() = 0;
+    virtual QByteArray config() = 0;
+    virtual QString iconPath(const QString &iconId) = 0;
```

Comment:
Those looks like they are `const` functions.

Reply:
I like to leave these choices to derived classes.

Reply:
>I like to leave these choices to derived classes.

What choices do you talk about?

Reply:
These methods are clearly not mutators so they should be declared as `const`.

Reply:
> What choices do you talk about?

we don't know the working of derived classes, it may happen this will force derive classes to use mutable that will break constness contract

> These methods are clearly not mutators so they should be declared as const.

what if they are caching the returned value?

Reply:
> > These methods are clearly not mutators so they should be declared as const.
> 
> what if they are caching the returned value?

Caching isn't mutating.

Reply:
> Caching isn't mutating.

well you're changing a internal variable (state) in a const method.

Reply:
> > Caching isn't mutating.
> 
> well you're changing a internal variable (state) in a const method.

You clearly misunderstand what is the state of the object.

Reply:
> You clearly misunderstand what is the state of the object.

one can't use QReadWriteLock with such a code structure.

Reply:
> > You clearly misunderstand what is the state of the object.
> 
> one can't use QReadWriteLock with such a code structure.

https://github.com/qbittorrent/qBittorrent/blob/master/src/base/settingsstorage.h
It isn't part of object state.

Reply:
> Not every internal variable is part of an object's state.

but you're still changing a variable in a const method, wouldn't that violate the constness contract?

Reply:
Constness is about externally visible state of object.
See a (quick) explanation of `mutable` specifier at cppreference.com: https://en.cppreference.com/w/cpp/language/cv

Reply:
but why mark a method const, if it's not *really* a const? what's the benefit? Users of the class can't use QReadWriteLock with such a class.

Reply:
> See a (quick) explanation of mutable specifier at cppreference.com

mutable is workaround constness contract, using it in itself violates the constness contract

Reply:
>but why mark a method const, if it's not really a const? 

It's really a const. It's just that you still don't understand what constness is.


> Users of the class can't use QReadWriteLock with such a class.

They can. I even gave you an example of a similar class above.

Reply:
> > See a (quick) explanation of mutable specifier at cppreference.com
> 
> mutable is workaround constness contract, using it in itself violates the constness contract

`mutable` isn't workaround as long as it is correctly used (i.e. class is correctly designed).

Reply:
> It's really a const. It's just that you still don't understand what constness is.

can you tell me what I don't understand?

>  I even gave you an example of a similar class above.

by implementing a mechanism of locking to manually preserve constness. Why give this extra work to classes that derive from an interface? The fewer assumptions in an interface the better.

Reply:
> Why give this extra work to classes that derive from an interface? The fewer assumptions in an interface the better.

Because interface and implementation are different things. Interface shouldn't make any assumptions about particular implementations.

Reply:
> Because interface and implementation are different things. Interface shouldn't make any assumptions about particular implementations.

This thread is about marking method const in an interface.

Reply:
>This thread is about marking method const in an interface.

Then add the `const` at the implementation site.

Reply:
> Then add the const at the implementation site if that is possible.

I don't think that's possible, regardless all member variables of all derived classes are const.

Reply:
From https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#con2-by-default-make-member-functions-const :
>A member function should be marked const unless it changes the object’s observable state.

If your interface does not have `const`, then the implementation is free to make changes to the *observable state*. That is, you permit that may happen, is that how an implementation of this interface really should behave?

Reply:
>A member function should be marked const unless it changes the object’s observable state.

This is not quite correct. But it seems that you often stick to it, so we have disagreements. It should be stated as follows:
>A member function should be marked `const` if it **isn't intended to change** the object’s observable state.

The problem is that it is often unclear what is the observable state of some types. For example, for types representing some value (e.g.  QString), this is quite obvious. But what is real state of UIThemeSource? IMO, it has no own properties so its observable state is immutable at all.

Reply:
>This is not quite correct. But it seems that you often stick to it, so we have disagreements. It should be stated as follows:

Just to be clear, almost every specifier in C++ is about intention. That quote is from the guidelines, not from me and you should consider submit a reword for it (to help others to not be misguided).

Reply:
>I suppose the guidelines means that it is easier (to write better code?) to assume a member function is const and relax it afterwards than doing it the other way around.

This is clearly not the way to write better code, IMO. At least it won't teach you to design types/interfaces correctly. Unfortunately C++ is "implementation-oriented" language. It doesn't make clear distinction between types/interfaces and their implementations.

Reply:
I am not experienced enough to comment about objects states and relations. My argument is simple, marking a method const signals to users that the method is thread-safe and can be used without external locking. The problem with marking a method const in an interface is that it makes assumptions about the implementation but I've learned that the fewer assumptions in an interface the better.

Reply:
>can you please tell me what I don't understand?

@jagannatharjun 
I think this would help you understand it better:
https://isocpp.org/wiki/faq/const-correctness#logical-vs-physical-const
Please read it before further discussions.

Offtopic, compared to "C++ Core Guidelines", "C++ Super-FAQ" looks better written.

--- 
Furthermore, now I think `QRCThemeSource` and the other one should read-in the file contents at the constructor instead of at access time. I feel it that way is probably because of the naming of the functions. If you want it read-in at access time, then I suggest renaming them to something like `readStyleSheet()` or `loadStyleSheet()` so it would reflect its true action.

Reply:
> marking a method const signals to users that the method is thread-safe and can be used without external locking.

It has nothing to do with thread safety.

> The problem with marking a method const in an interface is that it makes assumptions about the implementation

You shouldn't make any assumptions about implementation. Constness of methods is part of type interface.
E.g. `Human::name()` is intended to not change `Human` state independently from any particular `Human` implementation. And `Human::setName()` is intended to change `Human` state independently from any particular `Human` implementation. So `Human::setName()` should be non-const even if implementation stores data in external database and does not change any class variable when changing the name. But  `Human::name()` should be const even if implementation obtains the value from some external storage and cache it in class variable.

Reply:
changed